corehq/apps/userreports/tests/test_choice_provider.py

```diff
@@ -0,0 +1,305 @@
+
+    def _test_get_choices_for_values(self, values):
+        self.assertEqual(
+            self.choice_provider.get_choices_for_values(values),
+            self.static_choice_provider.get_choices_for_values(values),
+        )
+
+    def test_query_no_search_all(self):
+        self._test_query(ChoiceQueryContext('', limit=20, page=0))
+
+    def test_query_no_search_first_short_page(self):
+        self._test_query(ChoiceQueryContext('', 2, page=0))
+
+    def test_query_no_search_second_short_page(self):
+        self._test_query(ChoiceQueryContext('', 2, page=1))
+
+    @abstractmethod
+    def test_query_search(self):
+        """
+        Suggested implementation:
+
+            self._test_query(ChoiceQueryContext('relevant_search_term', 2, page=1))
+
+        """
+        pass
+
+    @abstractmethod
+    def test_get_choices_for_values(self):
+        """
+        Suggested implementation:
+
+            self._test_get_choices_for_values(
+                [irrelevant_value, relevant_value, relevant_value])
+        """
+        pass
+
+
+class LocationChoiceProviderTest(TestCase, ChoiceProviderTestMixin):
+    dependent_apps = [
+        'corehq.apps.commtrack', 'corehq.apps.locations', 'corehq.apps.products',
+        'custom.logistics', 'custom.ilsgateway', 'custom.ewsghana', 'corehq.couchapps'
+    ]
+    domain = 'location-choice-provider'
+
+    @classmethod
+    def make_location(cls, location_code, location_name, domain=None):
+        domain = domain or cls.domain
+        return make_loc(location_code, location_name, type='outlet', domain=domain)
+
+    @classmethod
+    def setUpClass(cls):
+        report = ReportConfiguration(domain=cls.domain)
+        bootstrap_location_types(cls.domain)
+
+        location_code_name_pairs = (
+            ('cambridge_ma', 'Cambridge'),
+            ('somerville_ma', 'Somerville'),
+            ('boston_ma', 'Boston'),
+        )
+        cls.locations = []
+        choices = []
+
+        for location_code, location_name in location_code_name_pairs:
+            location = cls.make_location(location_code, location_name)
+            cls.locations.append(location)
+            choices.append(SearchableChoice(location.location_id, location.sql_location.display_name,
+                                            searchable_text=[location_code, location_name]))
+        choices.sort(key=lambda choice: choice.display)
+        cls.choice_provider = LocationChoiceProvider(report, None)
+        cls.static_choice_provider = StaticChoiceProvider(choices)
+
+    @classmethod
+    def tearDownClass(cls):
+        delete_all_locations()
+
+    def test_query_search(self):
+        self._test_query(ChoiceQueryContext('e', 2, page=0))
+        self._test_query(ChoiceQueryContext('e', 2, page=1))
+
+    def test_get_choices_for_values(self):
+        self._test_get_choices_for_values(
+            ['made-up', self.locations[0].location_id, self.locations[1].location_id])
```

Comment:
I don't totally understand this (though I see it matches the suggested pattern above). The goal is to test that passing something irrelevant to this function returns the base config?